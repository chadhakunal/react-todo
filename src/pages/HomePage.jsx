import React from "react";
import AddTodo from "components/AddTodo";
import TodoSection from "components/TodoSection";
import Search from "components/Search";

class HomePage extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      todo: [],
      completed: [],
      searching: false,
      filteredTodo: [],
      filteredCompleted: []
    };
  }

  localStorageSaveTimeout = 3000;

  componentDidMount() {
    const todo = JSON.parse(localStorage.getItem("todo") || "[]");
    const completed = JSON.parse(localStorage.getItem("completed") || "[]");
    this.setState({ todo, completed });
  }

  handleAdd = ({ title, description, timestamp }) => {
    let todo = [...this.state.todo];
    todo.push({ title, description, timestamp });
    this.setState({ todo });
    requestIdleCallback( () => {
      localStorage.setItem("todo", JSON.stringify(todo));
    }, {timeout: this.localStorageSaveTimeout});
  }

  handleDelete = (index, type) => {
    let items = this.state[type].slice(index, 1);
    this.setState({ [type]: items })
    requestIdleCallback( () => {
      localStorage.setItem(type, JSON.stringify(items));
    }, {timeout: this.localStorageSaveTimeout});
  }

  handleItemEdit = (index, type, title, description) => {
    let items = [...this.state[type]];
    if (title) { items[index].title = title };
    if (description) { items[index].description = description };
    this.setState({ [type]: items });
    requestIdleCallback( () => {
      localStorage.setItem(type, JSON.stringify(items));
    }, {timeout: this.localStorageSaveTimeout});
  }

  toggleType = (type) => {
    return type === "todo" ? "completed" : "todo";
  }

  moveItem = (index, from) => {
    const to = this.toggleType(from);
    let fromItems = [...this.state[from]];
    let toItems = [...this.state[to]];

    let item = fromItems[index];
    fromItems.splice(index, 1);
    toItems.push(item);

    this.setState({ [from]: fromItems, [to]: toItems });
    requestIdleCallback( () => {
      localStorage.setItem(from, JSON.stringify(fromItems));
      localStorage.setItem(to, JSON.stringify(toItems));
    }, {timeout: this.localStorageSaveTimeout})
  }

  searchLogic = (data, text) => {
    const res = data.filter(d => { return d.title.toLowerCase().search(text.toLowerCase()) !== -1; })
    return res;
  }

  handleSearch = (text) => {
    if (text === null || text === "") {
      this.setState({ searching: false });
      return;
    }
    const { todo, completed } = { ...this.state };
    const filteredTodo = this.searchLogic(todo, text);
    let filteredCompleted = this.searchLogic(completed, text);
    const newState = (filteredTodo.length === 0 && filteredCompleted.length === 0 && !text) ? { searching: false } : { searching: true, filteredTodo: filteredTodo, filteredCompleted: filteredCompleted };
    this.setState({ ...newState });
  }

  render() {
    return (
      <div className="container-fluid">
        <div className="row w-100 justify-content-center mt-5">
          <div className="col-10">
            <AddTodo onAdd={this.handleAdd}></AddTodo>
          </div>
        </div>
        <div className="row w-100 justify-content-center mt-5">
          <div className="col-10">
            <Search className="w-100" onSearch={this.handleSearch}></Search>
          </div>
        </div>
        <div className="row w-100 justify-content-center mt-3">
          <div className="col-md-6">
            <TodoSection name="Todo" type="todo" items={this.state.searching ? this.state.filteredTodo : this.state.todo} onDelete={this.handleDelete} onEdit={this.handleItemEdit} moveItem={this.moveItem} />
          </div>
          <div className="col-md-6">
            <TodoSection name="Completed" type="completed" items={this.state.searching ? this.state.filteredCompleted : this.state.completed} onDelete={this.handleDelete} onEdit={this.handleItemEdit} moveItem={this.moveItem} />
          </div>
        </div>
      </div>
    )
  }
}

export default HomePage
