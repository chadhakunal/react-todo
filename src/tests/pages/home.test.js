import { shallow, mount } from 'enzyme';
import HomePage from 'pages/home';

describe('HomePage', () => {
    it('should render correctly with no props', () => {
        const component = shallow(< HomePage />);
        expect(component).toMatchSnapshot();
    });
    
    it('should update state from localStorage on mount', ()=>{
        const component = mount(< HomePage />)
    })

    it('should add an entry to todo and update localStorage', ()=>{

    })

    it("should remove the entry from items and update localStorage", ()=>{
    
    });
    
    it("should update the entry from items and update localStorage", ()=>{
    
    });

    it("should move the entry from todo to completed and update localStorage", ()=>{
    
    });

    it("should move the entry from completed to todo and update localStorage", ()=>{
    
    });

    it("should send filtered items if input entered in search", ()=>{

    })

    it("should send all items if no input is entered in search", ()=>{

    })
})