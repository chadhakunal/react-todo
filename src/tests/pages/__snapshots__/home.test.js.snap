// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should render correctly with no props 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row w-100 justify-content-center mt-5"
  >
    <div
      className="col-10"
    >
      <AddTodo
        onAdd={[Function]}
      />
    </div>
  </div>
  <div
    className="row w-100 justify-content-center mt-5"
  >
    <div
      className="col-10"
    >
      <Search
        className="w-100"
        onSearch={[Function]}
      />
    </div>
  </div>
  <div
    className="row w-100 justify-content-center mt-3"
  >
    <div
      className="col-md-6"
    >
      <TodoSection
        items={Array []}
        moveItem={[Function]}
        name="Todo"
        onDelete={[Function]}
        onEdit={[Function]}
        type="todo"
      />
    </div>
    <div
      className="col-md-6"
    >
      <TodoSection
        items={Array []}
        moveItem={[Function]}
        name="Completed"
        onDelete={[Function]}
        onEdit={[Function]}
        type="completed"
      />
    </div>
  </div>
</div>
`;
